### PR TITLE
Add studio agent implementation

### PR DESCRIPTION
**Pull request by robbertvanginkel**
Refs BSR-59. This will still need to be hooked up to a subcommand, that will be done in a follow up pr.

### CODE REVIEW

File: proto/buf/alpha/studio/v1alpha1/invoke.proto [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// services methods here.
+package buf.alpha.studio.v1alpha1;
+
+message Metadata {
+  string key = 1;
+  repeated string value = 2;
+}
+
+message InvokeRequest {
+  string target = 1;
```
Comment (cyinma) [inline proto/buf/alpha/studio/v1alpha1/invoke.proto:32]: can we have a comment for the `target` field? I think other fields are self-explained enough, but `target` probably needs more explanation

File: private/bufpkg/bufstudioagent/plain_post_handler.go [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+	"go.uber.org/zap"
+	"golang.org/x/net/http2"
+	"google.golang.org/protobuf/proto"
+)
+
+// maxMessageSizeBytesDefault determines the maximum number of bytes to read
+// from the request body.
+const maxMessageSizeBytesDefault = 1024 * 1024 * 5
+
+func newPlainPostHandler(
```
Comment (bufdev) [inline private/bufpkg/bufstudioagent/plain_post_handler.go:40]: constructor should always be below struct definition

File: proto/buf/alpha/studio/v1alpha1/invoke.proto (hunk 1 above)
Comment (bufdev) [reply proto/buf/alpha/studio/v1alpha1/invoke.proto:32]: can we have comments for everything really

File: proto/buf/alpha/studio/v1alpha1/invoke.proto [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+syntax = "proto3";
+
+// Because the request agent needs to be able to receive simple CORS requests,
+// we're limited to the text/plain Content-Type. This isn't suported out of the
+// box in gRPC-Web, so for now we avoid using it and do plain POST with base64
+// proto request/response messages.
+//
+// We may explore websocket transports later, at which point we should define
+// services methods here.
+package buf.alpha.studio.v1alpha1;
```
Comment (bufdev) [inline proto/buf/alpha/studio/v1alpha1/invoke.proto:24]: can this comment include what this package is for?

File: proto/buf/alpha/studio/v1alpha1/invoke.proto [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+message Metadata {
+  string key = 1;
+  repeated string value = 2;
+}
+
+message InvokeRequest {
```
Comment (bufdev) [inline proto/buf/alpha/studio/v1alpha1/invoke.proto:31]: what is this for? all needs comments

File: proto/buf/alpha/studio/v1alpha1/invoke.proto [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+// Because the request agent needs to be able to receive simple CORS requests,
+// we're limited to the text/plain Content-Type. This isn't suported out of the
+// box in gRPC-Web, so for now we avoid using it and do plain POST with base64
+// proto request/response messages.
+//
+// We may explore websocket transports later, at which point we should define
+// services methods here.
+package buf.alpha.studio.v1alpha1;
+
+message Metadata {
```
Comment (bufdev) [inline proto/buf/alpha/studio/v1alpha1/invoke.proto:26]: don't name protobuf message `*data` - this should likely be `Header`

File: private/bufpkg/bufstudioagent/buffer_codec.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+	connect "github.com/bufbuild/connect-go"
+	"google.golang.org/protobuf/proto"
+)
+
+// bufferCodec is a `connect.Codec` for use with clients of type
+// `connect.Client[bytes.Buffer, bytes.Buffer]` which does not attempt to parse
+// messages but instead allows the application layer to work on the buffers
+// directly. This is useful for creating proxies.
+type bufferCodec struct {
```
Comment (bufdev) [inline private/bufpkg/bufstudioagent/buffer_codec.go:30]: needs a constructor

File: private/bufpkg/bufstudioagent/buffer_codec.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	connect "github.com/bufbuild/connect-go"
+	"google.golang.org/protobuf/proto"
+)
+
+// bufferCodec is a `connect.Codec` for use with clients of type
```
Comment (bufdev) [inline private/bufpkg/bufstudioagent/buffer_codec.go:26]: Godoc isn't markdown, remove backticks

File: proto/buf/alpha/studio/v1alpha1/invoke.proto (hunk 3 above)
Comment (robbertvanginkel) [reply proto/buf/alpha/studio/v1alpha1/invoke.proto:40]: Added a better description of what these envelope messages are for.

File: private/bufpkg/bufstudioagent/buffer_codec.go (hunk 6 above)
Comment (robbertvanginkel) [reply private/bufpkg/bufstudioagent/buffer_codec.go:30]: We changed the style guide to not require constructors. An empty name here is valid for the connect codec interface so there also wouldn't be anything to check.

File: proto/buf/alpha/studio/v1alpha1/invoke.proto (hunk 5 above)
Comment (robbertvanginkel) [reply proto/buf/alpha/studio/v1alpha1/invoke.proto]: Renamed to header, although the gRPC and Connect protocols both refer to this concept as `Metadata`. What is problematic about proto message names ending in `*data`?

File: private/bufpkg/bufstudioagent/plain_post_handler.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	studiov1alpha1 "github.com/bufbuild/buf/private/gen/proto/go/buf/alpha/studio/v1alpha1"
+	"github.com/bufbuild/connect-go"
+	"go.uber.org/zap"
+	"golang.org/x/net/http2"
+	"google.golang.org/protobuf/proto"
+)
+
+// maxMessageSizeBytesDefault determines the maximum number of bytes to read
+// from the request body.
+const maxMessageSizeBytesDefault = 1024 * 1024 * 5
```
[the remark this answers is not in the document]
Comment (robbertvanginkel) [reply private/bufpkg/bufstudioagent/plain_post_handler.go:38]: The requests won't be device-local if sent to an in infra hosted agent beyond a SSO provider? 5MB incoming wire size is above the 4mb default unpacked gRPC message size, so not sure what orders of magnitude bigger would get you here?

With all the requests either originating from the web interface and returning there, having some default limit on a POST endpoint we expect people to expose seems like a reasonable failsafe. Could make this configurable, but would like to understand related to your q's above better?

**Review by lrewega — APPROVED** [review]
LGTM

**Pull request merged**